File: config/feature_flags.py
"""
Feature flags for staged/experimental code that has been merged to main but
is not yet meant to be active. A plain dict, checked explicitly at the call
site -- not a comment, not a commit-message date, not a docstring promise.

Why this exists: commit 7fc986d ("Staged: ... VID-QP fixes (hold for
08-22) (#40)", merged 2026-08-16) added a shadow-mode diagnostic to
generate_quiet_panic_video.py's run_all_gates() gated only by a comment
("SHADOW MODE (2026-08-16)... hold for 08-22"). Comments are not enforced
by CI and go live on merge regardless of wording -- the very next scheduled
run (Monday 2026-08-17) crashed on it. See CLAUDE.md's "Staged/experimental
code" rule.

Usage:
    from config.feature_flags import FEATURE_FLAGS
    if FEATURE_FLAGS.get("some_flag", False):
        ...

Flags default False. Flipping one to True is a deliberate, reviewable,
one-line diff -- not a side effect of an unrelated merge.
"""

FEATURE_FLAGS = {
    # VID-QP pre-TTS caption-length shadow-mode estimate (added 2026-08-16,
    # crashed 2026-08-17, shape bug fixed 2026-08-18 in
    # scripts/video/generate_quiet_panic_video.py's run_all_gates()).
    # Currently unused by that fix -- the shadow check itself always runs
    # and is always non-blocking by construction (see run_all_gates()'s
    # '_pre_tts_estimate' entry, which hardcodes 'pass': True regardless of
    # this flag). This flag is reserved for a future decision on whether to
    # promote the shadow check into a REAL blocking gate. Leave False until
    # that product decision is made explicitly.
    "qp_shadow_mode_estimate": False,

    # Cerebras fallback (VID-P4 engine.py, VID-QP quiet_panic_script_gen.py).
    # Added 2026-08-19: Cerebras has been payment-blocked (402 "Payment
    # required") since at least 2026-08-18 and Abhinav cannot add a payment
    # method to that account (see CLAUDE.md). Fallback order is now
    # Gemini -> Groq (free tier, 429 on limit rather than a permanent 402).
    # The Cerebras integration code is kept intact, not deleted, behind this
    # flag -- so it can be re-enabled with a one-line diff if billing is
    # ever resolved, without a rewrite. Leave False until Abhinav confirms
    # Cerebras billing is fixed.
    "cerebras_fallback_enabled": False,

    # VID-QP Groq fallback (quiet_panic_script_gen.py's _call_groq() --
    # VID-P4's own engine.py Groq fallback has its own separate flag pair
    # below, p4_groq_fallback_enabled/p4_groq_fallback_model, added in the
    # FINAL ARCHITECTURE PASS). Added 2026-08-22 (qwen rollout evidence
    # pass): Groq previously fired unconditionally the moment Gemini
    # failed, targeting the now-decommissioned llama-3.3-70b-versatile
    # (confirmed dead, live 404) -- effectively a silent no-op. Swapping to
    # a model that actually works (qwen/qwen3.6-27b, see the flag below)
    # behind the SAME unconditional call would put live qwen output into
    # production with zero review the moment this merges. This flag stops
    # that: Groq fallback is now skipped entirely (falls through to the
    # Cerebras check, or raises) until Abhinav explicitly enables it after
    # reviewing the rollout's evidence (5-set P4 comparison + QP voice
    # test). Leave False until that review happens.
    "qp_groq_fallback_enabled": False,

    # Which model quiet_panic_script_gen.py's _call_groq() targets, once
    # qp_groq_fallback_enabled above is True. qwen/qwen3.6-27b chosen over
    # the dead llama-3.3-70b-versatile after real Groq-side testing (fits
    # the trimmed codex's TPM budget, reasoning_effort='none' required --
    # see _call_groq()'s docstring). Kept as its own flag value (not
    # hardcoded in _call_groq()) so a future model swap or rollback is a
    # one-line config change, not a code change.
    "qp_groq_fallback_model": "qwen/qwen3.6-27b",

    # VID-P4 Groq fallback (engine.py's _try_groq()). Added 2026-08-22 as
    # part of the FINAL ARCHITECTURE PASS -- same reasoning as
    # qp_groq_fallback_enabled/qp_groq_fallback_model above, brought to
    # parity: _try_groq() previously fired unconditionally on Gemini
    # failure, targeting the dead llama-3.3-70b-versatile. Separate from
    # QP's flags because Phase A's evidence showed materially worse qwen
    # results against VID-P4's prompt (word-count overruns on 4/5 test
    # sets, bad price math on 2/5) -- P4 and QP must be independently
    # enable-able, not coupled to the same switch, so a decision on one
    # pipeline never silently drags the other along.
    "p4_groq_fallback_enabled": False,
    "p4_groq_fallback_model": "qwen/qwen3.6-27b",
}

File: src/lib/feature-flags.ts
/**
 * TypeScript mirror of config/feature_flags.py -- see that file's docstring
 * for the full rationale (commit 7fc986d incident, 2026-08-16/17). Kept as
 * a separate file rather than a shared JSON config to avoid adding a
 * cross-language build/import dependency for a two-flag object; keep the
 * two files' values in sync by hand when either changes.
 *
 * Usage:
 *   import { FEATURE_FLAGS } from '@/lib/feature-flags';
 *   if (FEATURE_FLAGS.cerebrasFallbackEnabled) { ... }
 */

export const FEATURE_FLAGS = {
  // Cerebras fallback (article pipeline's generate-with-failover.ts; see
  // also config/feature_flags.py's 'cerebras_fallback_enabled' for the
  // Python-side video pipelines). Added 2026-08-19: Cerebras has been
  // payment-blocked (402 "Payment required") since at least 2026-08-18 and
  // Abhinav cannot add a payment method to that account. Fallback order is
  // now Gemini -> Groq (free tier, 429 on limit rather than a permanent
  // 402). The Cerebras integration code is kept intact, not deleted,
  // behind this flag -- so it can be re-enabled with a one-line diff if
  // billing is ever resolved. Leave false until Abhinav confirms Cerebras
  // billing is fixed.
  cerebrasFallbackEnabled: false,

  // Article-pipeline Groq fallback (groq.ts / generate-with-failover.ts).
  // Added 2026-08-22 (qwen rollout evidence pass): Groq previously fired
  // unconditionally whenever Gemini failed and a Groq key was configured,
  // targeting the now-decommissioned llama-3.3-70b-versatile (confirmed
  // dead, live 404) -- meaning every retryable Gemini failure fell through
  // to a Groq call that always 404'd, then (since cerebrasFallbackEnabled
  // is also false) threw BothProvidersFailedError, which per the 2026-06-28
  // recycle policy gets the draft DELETED, not retried. Swapping to a
  // working model behind the SAME unconditional call would put live qwen
  // output into production with zero review the moment this merges. This
  // flag stops that: Groq fallback is skipped entirely until Abhinav
  // explicitly enables it after reviewing the rollout's evidence (real TPM
  // probe + 3-sample sanity check against src/lib/hard-rules.ts).
  //
  // Flipped true 2026-08-22 (FINAL ARCHITECTURE PASS) per explicit
  // instruction, evidence reviewed -- this also restores a working
  // fallback for the silent-delete bug described above. IMPORTANT CAVEAT,
  // confirmed live via `gh secret list`: GROQ_API_KEY is NOT currently a
  // GitHub repo secret -- it only exists in a local scripts/test/.env
  // (explicitly marked "local/test credential only" when first provided).
  // Flipping this flag is therefore currently a functional no-op in
  // production: GroqProvider's fetch will still be attempted, but every
  // real scheduled workflow run has an empty apiKey, so Groq will 401
  // immediately and generateWithFailover() falls through to
  // BothProvidersFailedError exactly as before. Real Groq calls (and the
  // silent-delete fix) will not take effect until Abhinav adds
  // GROQ_API_KEY as an actual repo secret -- deliberately not done here,
  // per the earlier explicit instruction not to add it without a
  // dedicated decision. scripts/canary/model_canary.py's daily run
  // reports this exact condition ("SECRET NOT CONFIGURED") until resolved.
  articleGroqFallbackEnabled: true,

  // Which model groq.ts's GroqProvider targets, once
  // articleGroqFallbackEnabled above is true. Swapped 2026-09-17: qwen/
  // qwen3.6-27b (chosen 2026-08-22) was itself decommissioned without a
  // formal deprecations-page entry -- Groq rotates its Preview-tier Qwen
  // point releases (3-32b -> 3.6-27b -> 3.8-27b) without the same notice
  // as GA models, so pinning to one is exactly what killed this fallback
  // twice now (first llama-3.3-70b-versatile, then qwen3.6-27b). Moved to
  // openai/gpt-oss-120b -- Groq's own recommended replacement for BOTH
  // prior deprecations (console.groq.com/docs/deprecations) and a GA
  // production model, not Preview. Confirmed live via a real Groq API
  // call before this change shipped (model_canary.py workflow_dispatch
  // run, not just docs). Kept as its own flag value, not hardcoded in
  // groq.ts, so a future model swap or rollback is a one-line config
  // change.
  articleGroqFallbackModel: 'openai/gpt-oss-120b',
} as const;
